src/vm/class.cpp

```diff
@@ -4024,43 +3770,528 @@ void EEClassLayoutInfo::ParseFieldNativeTypes(
+    }
+
+    if (cInstanceFields == 0 && classSizeInMetadata == 0 && (!fParentHasLayout || pParentLayoutInfo->IsZeroSized()))
+    {
+        pEEClassLayoutInfoOut->SetIsZeroSized(TRUE);
+    }
+
+    // Calculate the managedsequential layout if the type is eligible.
+    if (!fDisqualifyFromManagedSequential || fExplicitOffsets || isBlittable)
+    {
+        BYTE parentManagedAlignmentRequirement = 0;
+        UINT32 parentSize = 0;
+        if (pParentMT && (pParentMT->IsManagedSequential() || (pParentMT->GetClass()->HasExplicitFieldOffsetLayout() && pParentMT->IsBlittable())))
+        {
+            parentManagedAlignmentRequirement = pParentLayoutInfo->m_ManagedLargestAlignmentRequirementOfAllMembers;
+            parentSize = pParentMT->GetNumInstanceFieldBytes();
+        }
+
+        CalculateSizeAndFieldOffsets(
+            parentSize,
+            cInstanceFields,
+            fExplicitOffsets,
+            pSortArray,
+            classSizeInMetadata,
+            packingSize,
+            parentManagedAlignmentRequirement,
+            /*limitToMaxInteropSize*/ FALSE,
+            &pEEClassLayoutInfoOut->m_ManagedLargestAlignmentRequirementOfAllMembers,
+            &pEEClassLayoutInfoOut->m_cbManagedSize);
+
+        if (pEEClassLayoutInfoOut->m_cbManagedSize == 0)
+        {
+            _ASSERTE(pEEClassLayoutInfoOut->IsZeroSized());
+            pEEClassLayoutInfoOut->m_cbManagedSize = 1; // Bump the managed size of the structure up to 1.
+        }
+    }
+
+    pEEClassLayoutInfoOut->SetIsManagedSequential(!fDisqualifyFromManagedSequential);
+    return;
+}
+#ifdef _PREFAST_
+#pragma warning(pop)
+#endif // _PREFAST_
+
+void EEClassNativeLayoutInfo::InitializeNativeLayoutFieldMetadataThrowing(MethodTable* pMT)
+{
+    CONTRACTL
+    {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_ANY;
+        PRECONDITION(CheckPointer(pMT));
+        PRECONDITION(pMT->HasLayout());
+    }
+    CONTRACTL_END;
+
+    EEClass* pClass = pMT->GetClass();
+    EEClassLayoutInfo* pLayoutInfo = pClass->GetLayoutInfo();
+
+    if (pClass->GetNativeLayoutInfo() == nullptr)
+    {
+        ListLockHolder nativeTypeLoadLock(pMT->GetDomain()->GetNativeTypeLoadLock());
+        ListLockEntryHolder entry(ListLockEntry::Find(nativeTypeLoadLock, pMT->GetClass()));
+        ListLockEntryLockHolder pEntryLock(entry, FALSE);
+        nativeTypeLoadLock.Release();
+        if (!pEntryLock.DeadlockAwareAcquire())
+        {
+            DefineFullyQualifiedNameForClassW()
+            COMPlusThrow(kTypeLoadException, IDS_CANNOT_MARSHAL_RECURSIVE_DEF, GetFullyQualifiedNameForClassW(pMT));
+        }
+
+        if (pClass->GetNativeLayoutInfo() == nullptr)
+        {
+            EEClassNativeLayoutInfo* pNativeLayoutInfo;
+            CollectNativeLayoutFieldMetadataThrowing(pMT, &pNativeLayoutInfo);
+            ((LayoutEEClass*)pClass)->m_nativeLayoutInfo.SetValue(pNativeLayoutInfo);
+#ifdef FEATURE_HFA
+            pMT->GetClass()->CheckForNativeHFA();
```

Comment:
This does not look thread safe. I think all initialization of `pNativeLayoutInfo` needs to be done before `m_nativeLayoutInfo` is set to point to it.